subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/runner/TestRunnerImpl.kt

```diff
@@ -66,9 +67,27 @@ internal class TestRunnerImpl(
 
                     val expectedResultsCount = tests.count()
 
+                    val gottenResultsByDeviceName : MutableMap<String, MutableList<TestRunResult>> = mutableMapOf()
+                    val expectedResultsCountByDeviceName = tests
+                        .groupBy { it.deviceName.name }
+                        .map { (deviceName, tests) ->
+                            gottenResultsByDeviceName[deviceName] = mutableListOf()
+                            deviceName to tests.count()
+                        }
+
                     val gottenResults = mutableListOf<TestRunResult>()
                     for (result in state.results) {
                         gottenResults.add(result)
+                        gottenResultsByDeviceName[result.request.testCase.deviceName.name]?.add(result)
+
+                        expectedResultsCountByDeviceName.forEach { (deviceName, expectedCount) ->
+                            val actualCount = gottenResultsByDeviceName[deviceName]?.size ?: 0
+                            if (actualCount >= expectedCount) {
+                                result.result
+                                state.deviceSignals.send(Device.Signal.ReservationNotNeeded(deviceName))
```

Comment:
That is not a DeviceSignal. We should use new channel. For example testRunSignals. And create TestRequestsForDeviceIsOver(deviceName: DeviceName)